python-connectors (new)

ldom opened the pull request.

bbejeck commented on python-connectors/README.md:
```diff
@@ -0,0 +1,60 @@
+
+To create a connector you can use `curl` or you can use the Control Center UI. Since we already have a terminal open, let's use that.
+
+Run `curl localhost:8083/connectors -X POST --json @./json/connector-fields.json` to create a connector with the example 1 above.
```
I had to wait a couple of minutes for this to work - not a big deal but maybe put a line there saying you might have to re-try if it doesn't work at first.  I think it could be due to the fact that my mac is `amd64` and not `arm64` - I did get this warning starting `connect The requested image's platform (linux/arm64) does not match the detected host platform (linux/amd64/v2) and no specific platform was requested ` it also took a couple of minutes for the `test-topic-XXX` to show up - but everything worked as expected in the end.

bbejeck reviewed (commented):
Nice tutorial @ldom! I just have a couple of minor comments but otherwise this looks good!

bbejeck reviewed (approved):
LGTM! 

Hi @ldom thanks for the contribution.  I was able to get this running end-to-end.

One thing to consider is the connect image is still pulling ARM64 (I've got an Intel mac), but it ran fine - it just took a bit longer than I suspect it would otherwise
![Screen Shot 2024-02-05 at 12 17 29 PM](https://github.com/confluentinc/tutorials/assets/199238/49cb29bb-e435-45b8-84a4-c17204741cfa)

davetroiano commented on python-connectors/README.md:
```diff
@@ -0,0 +1,60 @@
+
+1. Open a terminal and go to the tutorial folder - python-connectors.
+2. Start the cluster: `docker-compose up -d`
+3. Wait for a few seconds and run: `docker-compose ps`. Everything container instance should be marked as running.  
```
```suggestion
3. Wait for a few seconds and run: `docker compose ps`. Everything container instance should be marked as running.  
```

davetroiano commented on python-connectors/README.md:
```diff
@@ -0,0 +1,60 @@
+
+From there, you have to create a connector in the Connect cluster.  
+
+There are 3 examples provided with this tutorial in the `json`folder. They all use a python connector and the last one also includes a python SMT.
```
```suggestion
There are 3 examples provided with this tutorial in the `json` folder. They all use a Python connector and the last one also includes a Python SMT.
```

davetroiano commented on python-connectors/README.md:
```diff
@@ -0,0 +1,60 @@
+
+## Configuration Properties and Technical Details
+
+When you submit a json file to create a connector, it includes some settings for the python connector.
```
```suggestion
When you submit a json file to create a connector, it includes some settings for the Python connector.
```

davetroiano commented on python-connectors/README.md:
```diff
@@ -0,0 +1,60 @@
+## Configuration Properties and Technical Details
+
+When you submit a json file to create a connector, it includes some settings for the python connector.
+The main one is the `entry.point` property that is telling the connector which python script to use and which method to call repeatedly.
```
```suggestion
The main one is the `entry.point` property that is telling the connector which Python script to use and which method to call repeatedly.
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+## Introduction 
+
+The interface from java to python is done using the pemja library (developed by Alibaba).
+
+https://github.com/alibaba/pemja
```
```suggestion
The interface from Java to Python is done using the [pemja](https://github.com/alibaba/pemja) library (developed by Alibaba).
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+https://github.com/alibaba/pemja
+
+How pemja works:
+- It starts a thread for the python interpreter in the JVM
+- It calls python code using Java to C interface (JNI) and a C python module
```
```suggestion
- It starts a thread for the Python interpreter in the JVM
- It calls Python code using Java to C interface (JNI) and a C Python module
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+Step 1. Add the Python Tools jar to the `CLASSPATH`.
+
+Step 2. Put your python scripts, including an optional `requirements.txt` in a directory (the scripts directory).
```
```suggestion
Step 2. Put your Python scripts, including an optional `requirements.txt` in a directory (the scripts directory).
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+### How it works
+
+👉 When initializing (during the `configure()` call of the java SMT SDK), the SMT builds a python virtual environment 
```
```suggestion
👉 When initializing (during the `configure()` call of the Java SMT SDK), the SMT builds a Python virtual environment 
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+👉 When initializing (during the `configure()` call of the java SMT SDK), the SMT builds a python virtual environment 
+and (pip) installs the libraries referenced in the `requirements.txt` file. Then, it calls the `init()` method of the 
+python script (if configured). It also installs the `pemja` library in the virtual environment.
+
+For each call of the `apply()` call of the java SMT SDK (the method calling the transformation), it calls the `entry.point` 
+python method and passes a dict with the Kafka record (see next section).
```
```suggestion
Python script (if configured). It also installs the `pemja` library in the virtual environment.

For each call of the `apply()` call of the java SMT SDK (the method calling the transformation), it calls the `entry.point` 
Python method and passes a dict with the Kafka record (see next section).
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+### Python script and method signatures
+
+For transforms chained to source connectors, the python method called for the transform has the following signature:
```
```suggestion
For transforms chained to source connectors, the Python method called for the transform has the following signature:
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+### Filtering out a message
+
+To filter out a message, the python transform can return `None`.
```
```suggestion
To filter out a message, the Python transform can return `None`.
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+- `<transform.prefix>.type` must be `io.confluent.pytools.PyConnectSmt`
+- `<transform.prefix>.scripts.dir`: the directory where the python scripts reside. 
+- `<transform.prefix>.working.dir`: optional, the directory where to build the python virtual environment. If not passed, the scripts directory will be used.
```
```suggestion
- `<transform.prefix>.working.dir`: optional, the directory where to build the Python virtual environment. If not passed, the scripts directory will be used.
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+- `<transform.prefix>.type` must be `io.confluent.pytools.PyConnectSmt`
+- `<transform.prefix>.scripts.dir`: the directory where the python scripts reside. 
+- `<transform.prefix>.working.dir`: optional, the directory where to build the python virtual environment. If not passed, the scripts directory will be used.
+- `<transform.prefix>.entry.point`: Python entry point for the transform. See details on python entry points below. 
+- `<transform.prefix>.init.method`: optional, method called once by the SMT when it initializes the transform.
+- `<transform.prefix>.private.settings`: String passed to the python script. Can be used to put settings in JSON format; eg. `"{\"conf1\":\"value1\", \"conf2\":\"value2\"}"`.
+- `<transform.prefix>.offline.installation.dir`: optional, the directory containing wheel/python packages for offline installation of the packages in the virtual environment.
```
```suggestion
- `<transform.prefix>.entry.point`: Python entry point for the transform. See details on Python entry points below. 
- `<transform.prefix>.init.method`: optional, method called once by the SMT when it initializes the transform.
- `<transform.prefix>.private.settings`: String passed to the python script. Can be used to put settings in JSON format; eg. `"{\"conf1\":\"value1\", \"conf2\":\"value2\"}"`.
- `<transform.prefix>.offline.installation.dir`: optional, the directory containing wheel / Python packages for offline installation of the packages in the virtual environment.
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+- `<transform.prefix>.private.settings`: String passed to the python script. Can be used to put settings in JSON format; eg. `"{\"conf1\":\"value1\", \"conf2\":\"value2\"}"`.
+- `<transform.prefix>.offline.installation.dir`: optional, the directory containing wheel/python packages for offline installation of the packages in the virtual environment.
+
+**Note on python entry points**
+
+The format for the python entry points is driven by the way pemja works and the organization of python scripts inside the user modules. 
```
```suggestion
**Note on Python entry points**

The format for the python entry points is driven by the way pemja works and the organization of Python scripts inside the user modules. 
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+### Notes/FAQ
+
+- How to provide packages for offline installation of the python environment? Put the wheel packages in a directory and provide it using `<transform.prefix>.offline.installation.dir`.
+- The python script cannot/shouldn't change the type of the key or of the value.
```
```suggestion
- The Python script cannot change the type of the key or of the value.
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+Step 1. Add the Python Tools jar to the `CLASSPATH`.
+
+Step 2. Put your python scripts, including an optional `requirements.txt` in a directory (the scripts directory).
```
```suggestion
Step 2. Put your Python scripts, including an optional `requirements.txt` in a directory (the scripts directory).
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+See https://docs.confluent.io/platform/current/connect/devguide.html for Kafka Connect development concepts.
+
+👉 When initializing the connector task, the connector builds a python virtual environment
+and (pip) installs the libraries referenced in the `requirements.txt` file. Then, it calls the `init()` method of the
+python script (if configured). It also installs the `pemja` library in the virtual environment.
+
+For each call of the `poll()` call of the java Source connector SDK, it calls the `entry.point`
+python method and expects a single Kafka record or a list of Kafka records (see next section).
```
```suggestion
👉 When initializing the connector task, the connector builds a Python virtual environment
and (pip) installs the libraries referenced in the `requirements.txt` file. Then, it calls the `init()` method of the
Python script (if configured). It also installs the `pemja` library in the virtual environment.

For each call of the `poll()` call of the Java Source connector SDK, it calls the `entry.point`
Python method and expects a single Kafka record or a list of Kafka records (see next section).
```

davetroiano commented on python-connectors/tech-details.md:
```diff
@@ -0,0 +1,284 @@
+
+The `poll()` method can return either a list of records (described below) or a single one. 
+
+Records are expressed as python dicts containing 2 keys: `key` and `value`.
```
```suggestion
Records are expressed as Python dicts containing 2 keys: `key` and `value`.
```

davetroiano commented on python-connectors/tech-details.md:
````diff
@@ -0,0 +1,284 @@
+
+```
+
+On the java side, after the `poll()` method has been called, `SourceRecord` objects are created and sent to the destination topic.
````
```suggestion
On the Java side, after the `poll()` method has been called, `SourceRecord` objects are created and sent to the destination topic.
```

davetroiano commented on python-connectors/tech-details.md:
````diff
@@ -0,0 +1,284 @@
+```
+
+On the java side, after the `poll()` method has been called, `SourceRecord` objects are created and sent to the destination topic.
+From the python data, java stores the corresponding basic types:
````
```suggestion
From the Python data, Java stores the corresponding basic types:
```

davetroiano reviewed (commented):
nice job @ldom - the examples worked well for me. by far the majority of the comments here are suggestions to capitalize Python and Java. I may not have gotten them all so suggest find / replace

ldom commented:
I have updated the various docs and moved the Gdoc slides into a PDF. The connect docker image is pulled from cp-server-connect-datagen so I cannot update it for now. Eventually I'd prefer to build a clean one from scratch but it's too time-consuming for now.

davetroiano reviewed (approved):
LGTM!
